libraries/chain/combined_database.cpp

```diff
@@ -0,0 +1,499 @@
+#include <eosio/chain/combined_database.hpp>
+#include <eosio/chain/kv_chainbase_objects.hpp>
+
+namespace eosio { namespace chain {
+
+   combined_session::combined_session(chainbase::database& cb_database)
+       : type{ session_type::chainbase_only }, cb_session{ std::make_unique<chainbase::database::session>(
+                                                     cb_database.start_undo_session(true)) } {}
+
+   combined_session::combined_session(chainbase::database& cb_database, b1::chain_kv::undo_stack& kv_undo_stack)
+       : type{ session_type::chainbase_and_chain_kv }, kv_undo_stack{ &kv_undo_stack } {
+      try {
+         try {
+            cb_session = std::make_unique<chainbase::database::session>(cb_database.start_undo_session(true));
+            kv_undo_stack.push(false);
+         }
+         FC_LOG_AND_RETHROW()
+      }
+      CATCH_AND_EXIT_DB_FAILURE()
+   }
+
+   combined_session::combined_session(combined_session&& src) noexcept
+       : type(src.type), cb_session(std::move(src.cb_session)), kv_undo_stack(src.kv_undo_stack) {
+      src.type          = session_type::no_op;
+      src.kv_undo_stack = nullptr;
+   }
+
+   combined_session& combined_session::operator=(combined_session&& src) noexcept {
+      if (this != &src) {
+         type              = src.type;
+         cb_session        = std::move(src.cb_session);
+         kv_undo_stack     = src.kv_undo_stack;
+         src.type          = session_type::no_op;
+         src.kv_undo_stack = nullptr;
+      }
+      return *this;
+   }
+
+   void combined_session::push() {
+      switch (type) {
```

Comment:
When session_type is `chainbase_only`, push/squash/undo should not look like the `chainbase_and_chain_kv` path because that path terminates the application in case of any exceptions. That would change the current behavior of nodeos, which is to let exceptions through. Additionally, session_type is not boolean. It has three possible values. The `no_op` value allowed `combined_session` to absorb both `maybe_session` and `optional<combined_session>` elsewhere in the code.

Reply:
Why are we changing the behavior of cb_session for chainbase_and_chain_kv?  I want to make sure @b1bart and @larryk85 agree.  The same issue was brought up in Lin's PR, but I am not sure if it ever was resolved.

And for no_op is it any different than not processing because of 2 nullptrs, aside from having to check the nullptrs?